test/functional/tests/ended.js

```diff
@@ -2,90 +2,45 @@
 
-    // Test constants
-    var PLAYING_TIMEOUT = 10; // Timeout (in sec.) for checking playing status
-    var SEEK_SHIFT = 10; // Timeout (in sec.) for checking playback progress
-    var SEEK_TIMEOUT = 10; // Timeout (in sec.) for checking playback progress
-    var ENDED_TIMEOUT = SEEK_SHIFT + 10; // Timeout (in sec.) for checking seek to be completed
+// Suite name
+var NAME = 'ENDED';
 
-    var load = function(stream) {
-        registerSuite({
-            name: utils.testName(NAME, stream),
+exports.register = function(stream) {
 
-            load: function() {
-                if (!stream.available) this.skip();
-                if (stream.dynamic) this.skip();
-                utils.log(NAME, 'Setup');
-                command = this.remote.get(require.toUrl(intern.config.testPage));
-                return command.execute(player.loadStream, [stream])
-                .then(function() {
-                    // Check if playing
-                    utils.log(NAME, 'Check if playing');
-                    return command.executeAsync(player.isPlaying, [PLAYING_TIMEOUT]);
-                })
-                .then(function(playing) {
-                    assert.isTrue(playing);
-                });
-            }
-        })
-    };
+    suite(utils.testName(NAME, stream), (suite) => {
 
-    var seek = function(stream) {
-        registerSuite({
-            name: utils.testName(NAME, stream),
-
-            seek: function() {
-                if (!stream.available) this.skip();
-                if (stream.dynamic) this.skip();
-                // Seek the player before end
-                utils.log(NAME, 'Seek before end: ' + (stream.duration - SEEK_SHIFT));
-                return command.executeAsync(player.seek, [(stream.duration - SEEK_SHIFT), SEEK_TIMEOUT])
-                .then(function(seeked) {
-                    assert.isTrue(seeked);
-                });
-            }
+        before(async ({ remote }) => {
+            if (!stream.available || stream.dynamic) suite.skip();
+            utils.log(NAME, 'Load stream');
+            command = remote.get(intern.config.testPage);
+            await command.execute(player.loadStream, [stream]);
+            await command.executeAsync(player.isPlaying, [constants.EVENT_TIMEOUT]);
         });
-    }
-
-    var ended = function(stream) {
-        registerSuite({
-            name: utils.testName(NAME, stream),
 
-            ended: function() {
-                if (!stream.available) this.skip();
-                if (stream.dynamic) this.skip();
-                // Wait for 'ended' event
-                return command.executeAsync(player.waitForEvent, ['playbackEnded', ENDED_TIMEOUT])
-                .then(function(ended) {
-                    assert.isTrue(ended);
-                });
-            }
+        test('seek', async () => {
+            // Seek the player before end
+            let seekTime = (stream.duration - constants.SEEK_END_SHIFT);
```

Comment:
In rare cases duration might be < than seek_end_shift. I suggest we add `Math.max(0,value)`

Reply:
ok, agree